ostd/src/device/dispatcher/io_mem.rs

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! I/O Memory access dispatcher.
+
+use alloc::vec::Vec;
+use core::ops::Range;
+
+use log::debug;
+use spin::Once;
+
+use crate::{
+    device::io_mem::IoMem,
+    mm::{CachePolicy, PageFlags},
+    util::vaddr_alloc::VirtAddrAllocator,
+};
+
+/// I/O memory dispatcher that dispatches memory I/O access to device drivers.
+pub struct IoMemDispatcher {
+    allocators: Vec<VirtAddrAllocator>,
+}
+
+impl IoMemDispatcher {
+    /// Gets the I/O memory access for `range`.
+    ///
+    /// If the range is not available, then the return value will be `None`.
+    pub fn get(&self, range: Range<usize>) -> Option<IoMem> {
+        find_allocator(&self.allocators, &range)?
+            .alloc_specific(range.clone())
+            .ok()?;
+
+        debug!("Getting MMIO range:{:x?}..{:x?}", range.start, range.end);
+
+        // SAFETY: The created `IoMem` is guaranteed not to access physical memory or system device I/O.
+        unsafe { Some(IoMem::new(range, PageFlags::RW, CachePolicy::Uncacheable)) }
+    }
+
+    /// Recycles an MMIO range.
+    #[allow(dead_code)]
+    pub(crate) fn recycle(&self, range: Range<usize>) {
+        let Some(allocator) = find_allocator(&self.allocators, &range) else {
+            return;
+        };
+        debug!("Recycling MMIO range:{:x}..{:x}", range.start, range.end);
```

Comment:
There may be a blank line before this `debug!` line because you add a blank line before the `debug!` in `IoMemDispatcher::get`.